src/main/java/com/microsoft/azure/spring/data/documentdb/repository/query/DocumentDbQueryCreator.java

```diff
@@ -7,23 +7,57 @@
 
 import com.microsoft.azure.spring.data.documentdb.core.mapping.DocumentDbPersistentProperty;
 import com.microsoft.azure.spring.data.documentdb.core.query.Criteria;
+import com.microsoft.azure.spring.data.documentdb.core.query.Criteria.CriteriaType;
 import com.microsoft.azure.spring.data.documentdb.core.query.Query;
-import org.apache.commons.lang3.NotImplementedException;
+import org.javatuples.Pair;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.data.domain.Sort;
-import org.springframework.data.mapping.PropertyPath;
 import org.springframework.data.mapping.context.MappingContext;
-import org.springframework.data.mapping.context.PersistentPropertyPath;
 import org.springframework.data.repository.query.parser.AbstractQueryCreator;
 import org.springframework.data.repository.query.parser.Part;
 import org.springframework.data.repository.query.parser.PartTree;
 
 import java.util.Iterator;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 
 public class DocumentDbQueryCreator extends AbstractQueryCreator<Query, Criteria> {
 
+    private static final Logger logger = LoggerFactory.getLogger(DocumentDbQueryCreator.class);
+
+    private static final Map<Part.Type, Pair<CriteriaType, Boolean>> criteriaLookup;
+    static {
+        final Map<Part.Type, Pair<CriteriaType, Boolean>> init = new HashMap<>();
+        init.put(Part.Type.BEFORE, new Pair<>(CriteriaType.IS_LESS_THAN, Boolean.FALSE));
+        init.put(Part.Type.AFTER, new Pair<>(CriteriaType.IS_LESS_THAN_OR_EQUAL, Boolean.TRUE));
+        init.put(Part.Type.GREATER_THAN, new Pair<>(CriteriaType.IS_GREATER_THAN_OR_EQUAL, Boolean.FALSE));
+        init.put(Part.Type.GREATER_THAN_EQUAL, new Pair<>(CriteriaType.IS_GREATER_THAN, Boolean.FALSE));
```

Comment:
The boolean indicates if the logic is to be inverted i.e. NOT EQUAL or NOT NULL
I think I have some of the values wrong in there... they should be as follows
From DocumentDbQueryCreator
```java
        init.put(Part.Type.BEFORE, new Pair<>(CriteriaType.IS_LESS_THAN, Boolean.FALSE));
        init.put(Part.Type.AFTER, new Pair<>(CriteriaType.IS_GREATER_THAN, Boolean.FALSE));
        init.put(Part.Type.GREATER_THAN, new Pair<>(CriteriaType.IS_GREATER_THAN, Boolean.FALSE));
        init.put(Part.Type.GREATER_THAN_EQUAL, new Pair<>(CriteriaType.IS_GREATER_THAN_OR_EQUAL, Boolean.FALSE));
        init.put(Part.Type.LESS_THAN, new Pair<>(CriteriaType.IS_LESS_THAN, Boolean.FALSE));
        init.put(Part.Type.LESS_THAN_EQUAL, new Pair<>(CriteriaType.IS_LESS_THAN_OR_EQUAL, Boolean.FALSE));
```

From DocumentDbTemplate
```java
   private static final Map<CriteriaType, String> operatorLookup;
    static {
        final Map<CriteriaType, String> init = new HashMap<>();
        init.put(CriteriaType.IS_EQUAL,                 "r.@?=@@");
        init.put(CriteriaType.IS_LESS_THAN,             "r.@?<@@");
        init.put(CriteriaType.IS_LESS_THAN_OR_EQUAL,    "r.@?<=@@");
        init.put(CriteriaType.IS_GREATER_THAN,          "r.@?>@@");
        init.put(CriteriaType.IS_GREATER_THAN_OR_EQUAL, "r.@?>=@@");
```

Reply:
Thinking about it , would it be clearer to eliminate the boolean in the criteriaLookup and extend the map in DocumentDbQueryCreator to support all the criteria?

Reply:
Personally I think it will be clearer to eliminate the boolean.

One reason is that it's hard to tell what does the boolean mean here without comment.

Another reason is that it's easy to make mistake and hard to detect/fix the mistake, and for other methods which use this Map, they have to use the Map with much carefulness.

Reply:
I'll make the changes later today push a commit along with a few other alterations based on the feedback.

Thanks again